### PR TITLE
IGNITE-13060-tracing-engine

### PR DESCRIPTION
**Pull request by sanpwc**
Tracing configuration engine + communication, exchange and discovery tracing.

### CODE REVIEW

File: examples/src/main/java/org/apache/ignite/examples/opencensus/OpenCensusMetricsExporterExample.java [hunk 1]
```diff
@@ -31,7 +31,7 @@
 import org.apache.ignite.IgniteCache;
 import org.apache.ignite.Ignition;
 import org.apache.ignite.configuration.IgniteConfiguration;
-import org.apache.ignite.spi.metric.opencensus.OpenCensusMetricExporterSpi;
+import org.apache.ignite.opencensus.spi.metric.OpenCensusMetricExporterSpi;
```
Comment (nizhikov) [inline examples/src/main/java/org/apache/ignite/examples/opencensus/OpenCensusMetricsExporterExample.java:34]: Why this change?

File: modules/core/src/main/java/org/apache/ignite/IgniteSystemProperties.java [hunk 2]
```diff
@@ -17,13 +17,13 @@
 
 package org.apache.ignite;
 
+import javax.net.ssl.HostnameVerifier;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/IgniteSystemProperties.java:20]: Why this change?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/tracing/SpanManager.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+package org.apache.ignite.internal.processors.tracing;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Manager for {@link Span} instances.
+ */
+public interface SpanManager {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/tracing/SpanManager.java:26]: Why do we need two interface `SpanManger` and `Tracing`?
It seems for me that we can go with the `Tracing` and remove `SpanManager`.

File: modules/core/src/main/java/org/apache/ignite/spi/tracing/SpanStatus.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.spi.tracing;
+
+/**
+ * Various statuses for Spans execution.
+ */
+public enum SpanStatus {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/tracing/SpanStatus.java:23]: Can you, please, clarify, why do we need this statuses?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/tracing/SpanManager.java (hunk 3 above)
Comment (sanpwc) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/tracing/SpanManager.java:26]: Tracing is wider abstraction, that also include configuration. In "next" PR https://github.com/apache/ignite/pull/7974 it'll be updated to following state:
```
/**
 * Tracing sub-system interface.
 */
public interface Tracing extends SpanManager {
    /**
     * @return Helper to handle traceable messages.
     */
    public TraceableMessagesHandler messages();

    /**
     * Returns the {@link TracingConfigurationManager} instance that allows to
     * <ul>
     *     <li>Configure tracing parameters such as sampling rate for the specific tracing coordinates
     *          such as scope and label.</li>
     *     <li>Retrieve the most specific tracing parameters for the specified tracing coordinates (scope and label)</li>
     *     <li>Restore the tracing parameters for the specified tracing coordinates to the default.</li>
     *     <li>List all pairs of tracing configuration coordinates and tracing configuration parameters.</li>
     * </ul>
     * @return {@link TracingConfigurationManager} instance.
     */
    public @NotNull TracingConfigurationManager configuration();
}

```

File: modules/core/src/main/java/org/apache/ignite/spi/tracing/SpanStatus.java (hunk 4 above)
Comment (sanpwc) [reply modules/core/src/main/java/org/apache/ignite/spi/tracing/SpanStatus.java:23]: It's the concept, that we've borrowed from OpenCensus.

> Status defines a logical error model to represent a filterable state of the span at a point in time.
> Status allows tracing visualization tools to highlight unsuccessful spans and helps in debugging errors.

https://opencensus.io/tracing/span/status/
In Ignite statuses are used, for example, within discovery handling.
```
            if (locNodeId.equals(node.id())) {
                if (log.isDebugEnabled())
                    log.debug("Received join request for local node, dropping: " + msg);

                msg.spanContainer().span()
                    .addLog(() -> "Dropped")
                    .setStatus(SpanStatus.ABORTED)
                    .end();

                return;
            }
```
Besides other advantages statuses are used for automatic markup in trace visualization tools.

File: examples/src/main/java/org/apache/ignite/examples/opencensus/OpenCensusMetricsExporterExample.java (hunk 1 above)
Comment (sanpwc) [reply examples/src/main/java/org/apache/ignite/examples/opencensus/OpenCensusMetricsExporterExample.java:34]: Cause besides metrics tracing was added as one of OpenCensus related spis. So at the moment package tree is following:
```
-org.apache.ignite
  -opencensus.spi
    -metric
    -tracing
```

File: modules/core/src/main/java/org/apache/ignite/IgniteSystemProperties.java (hunk 2 above)
Comment (sanpwc) [reply modules/core/src/main/java/org/apache/ignite/IgniteSystemProperties.java:20]: No reason. Fixed.

File: examples/src/main/java/org/apache/ignite/examples/opencensus/OpenCensusMetricsExporterExample.java (hunk 1 above)
Comment (nizhikov) [reply examples/src/main/java/org/apache/ignite/examples/opencensus/OpenCensusMetricsExporterExample.java:34]: This clashes with the existing package hierarchy.
We place spi's to `org.apache.ignite.spi.SPIPACKAGE`.

Let's keep current naming hierarchy.

Comment (sanpwc) [reply examples/src/main/java/org/apache/ignite/examples/opencensus/OpenCensusMetricsExporterExample.java]: Should be fixed by now.

**Comment by dspavlov** [comment]
This PR is associated with a resolved ticket. Closing the PR, please reopen it if you still need it.
